tools/cldr-code/src/main/java/org/unicode/cldr/util/BoundaryTransform.java

```diff
@@ -0,0 +1,271 @@
+                toReplaceBy = toReplaceBy.replace("${" + namedGroup + "}", matchAfter.group(namedGroup));
+            }
+
+            return s.substring(0, hasB ? matchBefore.start("b") : boundary)
+                + toReplaceBy
+                + s.substring(hasA ? matchAfter.end("a") : boundary);
+        }
+        @Override
+        public String toString() {
+            return beforeBoundary.toString() + BOUNDARY + afterBoundary + START_REPLACE_BY + replaceBy;
+        }
+    }
+
+    private List<BoundaryRule> rules;
+
+    private BoundaryTransform(List<BoundaryRule> rules) {
+        this.rules = rules;
+    }
+
+    /**
+     * The format is:
+     * ruleset = rule (';' rule)*
+     * rule = before '❙' after '→' replaceBy
+     * before = (contextBefore '⦅')? replaceBefore
+     * after = replaceAfter ('⦆' contextAfter)?
+     * @param rulesPattern
+     * @return
+     */
+    public static BoundaryTransform from(String rulesPattern) {
+        List<BoundaryRule> rules = new ArrayList<>();
+        for (String rulePattern : SEMI.split(rulesPattern)) {
+            rules.add(new BoundaryRule(rulePattern));
+        }
+        return new BoundaryTransform(rules);
+    }
+
+    @Override
+    public String apply(String t, Integer u) {
+        int boundary = u;
+        for (BoundaryRule rule : rules) {
+            String result = rule.matches(t, 0, boundary, t.length());
+            if (result != null) {
+                return result;
+            }
+        }
+        return t;
+    }
+    @Override
+    public String toString() {
+        return JOIN_SEMI.join(rules);
+    }
+
+    // rule = contextBefore ⦅ replaceBefore ❙ replaceAfter ⦆ contextAfter → replaceBy
+
+    // TODO build with data
+    // TODO consider for general having ❙, ❴, or ❵. That is:
+    //   ❴ is only applied at the start of a placeholder
+    //   ❵ is only applied at the end of a placeholder
+    //   ❙ is applied at both positions
+
+
+    public enum BoundaryUsage {general, unitPrefixes}
+
+    private static Map<String, BoundaryTransform> GENERAL_LOCALE_TO_BOUNDARY_TRANSFORM = ImmutableMap.
+        <String, BoundaryTransform>builder()
+        .put("it", BoundaryTransform.from(""
+            + " e⦅ ❙⦆[eE]($|[^dD])→d ;" // ed before e (with exception)
+            + " o⦅ ❙⦆[oO]($|[^dD])→d ;" // od before o (with exception)
+            ))
+        .put("es", BoundaryTransform.from(""
+            + " ⦅y ❙⦆([iI]|[hH][iI]($|[^aAeE]))→e ;" // y
+            + " ⦅o ❙⦆([uU8]|[hH][oU]|11($| ))→u ;" // o
+            ))
+        .put("he", BoundaryTransform.from(""
+            + " \\u05D5⦅❙⦆\\P{sc=Hebr}→-;" // y
+            ))
+        .put("en", BoundaryTransform.from(
+            " a⦅ ❙⦆[aeiou]→n ;" // doesn't handle "an hour", or "an underground"
+            ))
+        .put("zh", BoundaryTransform.from(
+            "[\\p{L}&&\\p{sc=hani}]⦅❙⦆[\\p{L}&&\\P{sc=hani}]→ ;" // add space at chinese/non-chinese letter junction
```

Comment:
won't this add a space between Han chars and shared CJK punctuation like brackets etc?

Reply:
The \\p{L}&& restricts it to letters